### ISSUE TITLE
Can we optimize away an unused super class?

### ISSUE THREAD

**Issue by jabraham17** [chapel-lang/chapel#26051]
I came to the unfortunate realization while working on the binarytrees benchmark that not matter what all Chapel classes have as their first element a "super" field. This means that no matter what the code is we are always paying the cost of having that field there.

For example, this linked list node is always going to be 16 bytes.
```chapel
class Node {
  var next: unmanaged Node?;
}
```
This is because all Chapel classes "pay the cost" of a super class, even without using those features. The "Node" class is codegened approximately as:
```c
struct Root {
 int32_t cid; // 4 bytes
}; // 4 bytes
struct Node {
  struct Root super;  // 4 bytes
                                 // 4 bytes of padding for alignment
  struct Node* next; // 8 bytes (on 64-bit systems)
}; // 16 bytes
```

Why do we have to pay the cost of having this super field? I think today we already don't pay the time complexity cost of having this field, but we do pay the memory complexity cost?

Note that I think we don't pay the time complexity cost (at least for the LLVM backend, not sure about C but I assume/hope its true as well) because the backend is optimizing it away. The code as dumped by the Chapel compiler allocates the class and initializes the super class by setting the constant class id to the "cid" field. Then the backend is smart enough to see that the "cid" is never used and optimizes it away (yay DSE). But semantically the backend cannot change the memory layout of the types, so we just have empty uninitialized memory that is never used.

Can we instead write a compiler optimization that removes the "super" field/ "cid" field (I am lazily assuming they are interchangable here) that checks if they are never used, and in that case does not codegen that part of the class?

I believe this should be semantically valid, as AFAIK we make no guarantees to the user about the memory layout of classes.

I think this can be achieved by just checking if we ever call "getcid" or "testcid" on a given type, although this is perhaps a naive simplification.

**Comment by mppf** [comment]
I would expect that the reason this exists is that `delete`ing a class instance is implemented as methods on `RootClass` (née `Object`). In principle we could make that generic, but that approach wouldn't work for something like `list(owned RootClass)`.

In other words, my conclusion here is that this isn't practical as long as we support `RootClass`.

Note that for something like the binary trees benchmark, it might be better to use `c_ptr` or similar, which would trivially allow you to avoid such overheads.

**Comment by bradcray** [comment]
> Note that for something like the binary trees benchmark, it might be better to use c_ptr or similar, which would trivially allow you to avoid such overheads.

I'm not a fan of relying on `c_ptr` for a Chapel CLBG entry.

**Comment by bradcray** [comment]
> In the past we have talked about class Box { var x; } that would allow one to heap-allocate a record.

That kind of approach makes more sense to me for a Chapel CLBG entry, though I'd also wonder about other approaches.  But having regretted the need to pay the cost of a cid field for a completely flat class hierarchy that didn't need it at times, Jade's proposal also resonates with me.

A stopgap approach would be to have an attribute in the language that says to squash the cid field and then have the compiler complain if something requires it in a class for which it's been squashed.  This feels like a good use of attributes in the sense of communicating something to the compiler to get a change in implementation without a change in behavior.  If an end-user were requesting the feature, that's what I would propose, at least as a hopefully simple/potentially short-term fix.  I don't feel as wild about relying on it in a CLBG entry, though, but I'm also not dead-set against it.

**Comment by jabraham17** [comment]
While the binarytrees benchmark is what got me thinking about this, I agree that this has more utility beyond that. It made me think about C++  classes, which only have the virtual table pointer if you request it (via the virtual keyword), hence this issue.

In my opinion, a “Box” type to have a heap allocated record would only be a stopgap for this issue, and feels like a confusing usage of “class”. I like Brad’s suggestion of an attribute that removes the “cid” because it would not only allow us to have this optimization but the Box type could be trivially implemented (perhaps my naive assumption here). 

But I would take it a step further, if we have the ability to annotate a class to say “do not use a cid field” and we have the ability to check that a user has not misapplied that attribute, why can’t that attribute be automatically used in all applicable situations. I am thinking of this like Aggregators and the auto-aggregation optimization, users can force it by hand but in many cases we can automatically apply it.

**Comment by bradcray** [comment]
>  if we have the ability to annotate a class to say “do not use a cid field” and we have the ability to check that a user has not misapplied that attribute, why can’t that attribute be automatically used in all applicable situations. 

In proposing it, I was imagining that verification might be simpler than proving it's unused due to conservatism, but could definitely be wrong in that.

Originally, I was thinking that the analysis required here might be along the lines of "Has no subclasses" / "Has no overload methods", but Michael's "list of RootClass" case got me worried about that.  And I think without some sort of `final` keyword, we couldn't really determine that a class has no subclasses without a whole-program analysis?

**Comment by mppf** [comment]
I agree that an attribute might be interesting here.

What is unclear to me is if this offers enough performance improvement to be worth bothering with. Especially if it is hard to implement (such as if it requires a whole-program optimization).

**Comment by jabraham17** [comment]
> What is unclear to me is if this offers enough performance improvement to be worth bothering with. Especially if it is hard to implement (such as if it requires a whole-program optimization).

I agree it probably needs to be a whole-program optimization to work correctly, but it should be easier to evaluate the performance by doing the optimization by hand or even adding the attribute (for an experiment we don't need to validate correctness, just trust the user). Then we can decide if the perf benefits are worth the effort to do the checking properly.

FWIW, using the binary trees as an example benchmark I can modify the generated C code and recompile it to simulate the optimization.

On my Mac I get about a 1.3x speedup
- stock binarytrees-pool: 1.034s
- modified binarytrees-pool: 0.798s

Where the modification I made was to remove "super" from `chpl_Tree_chpl_object` and remove the lines to initialize `chpl__cid` for that "super" field.

This is a comparatively small benchmark, so it would be interesting to try that on a larger application, ideally something like Arkouda (if that is even tractable to manually edit all that generated code?)

**Comment by bradcray** [comment]
Beyond any time savings, it seems as though it could also reduce our pretty massive memory footprint by ~1/3 (?):  Read down the `mem` column at:  https://benchmarksgame-team.pages.debian.net/benchmarksgame/performance/binarytrees.html

**Comment by mppf** [comment]
Actually we already have something close to what the attribute would do, in the form of pragmas:

``` chapel
pragma "no object"
pragma "no default functions" // avoid errors about serialize
class MyClass {
  var x: int;
  proc deinit() {
    writeln("in MyClass.deinit");
  }
}

// to work around compilation errors
inline operator ==(a: borrowed MyClass?, b: _nilType) {
  return __primitive("ptr_eq", a, nil);
}
inline operator !=(a: borrowed MyClass?, b: _nilType) {
  return __primitive("ptr_neq", a, nil);
}
 
var x = new MyClass(1);
writeln(x.x);
```

Seems like it might serve for experimentation (but certainly wouldn't be nice in CLBG).
